### PR TITLE
feat: add charge name and description input validation

### PR DESCRIPTION
**Pull request by Mech0z**
## Description

Add NullOrWhitespace input validation of Charge Name and Charge Description for Charge Information Requests

## References

* #1326

### CODE REVIEW

File: docs/business-processes/asynchronous-validations.md [hunk 1]
```diff
@@ -36,6 +36,8 @@ The following asynchronous validation rules are currently implemented in the cha
 |VR.906|Transaction not completed: The request received contained multiple transactions for the same charge, and one of the previous transactions failed validation why this transaction is also rejected|D14|All|N/A|
 |VR.907|After creating a Charge, it is no longer allowed to change the resolution of that charge|D23|All|N/A|
 |VR.909|The number of prices received does not match the expected number of prices given the time interval and resolution provided.|E87|All|N/A|
+|VR.XXX|The Charge Name of a charge is mandatory|E0H|All|N/A|
+|VR.XXX|The Charge Description of a charge is mandatory|E0H|All|N/A|
```
Comment (prtandrup) [inline docs/business-processes/asynchronous-validations.md:40]:
```suggestion
|VR.XXX|Charge name is mandatory|E0H|All|N/A|
|VR.XXX|Charge description is mandatory|E0H|All|N/A|
```

At some point, we need to split the Charge types column into Charge info and Charge Prices, probably too soon right now.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/ChargeCommandNullChecker.cs [hunk 2]
```diff
@@ -50,12 +50,6 @@ private static void CheckChargeCommand(ChargeCommand chargeCommand)
         private static void CheckChargeOperation(ChargeOperationDto chargeOperationDto, BusinessReasonCode businessReasonCode)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/ChargeCommandNullChecker.cs:50]: `BusinessReasonCode ` can be removed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeDescriptionRequiredRule.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    public class ChargeDescriptionRequiredRule : IValidationRule
+    {
+        private readonly ChargeOperationDto _chargeOperationDto;
+
+        public ChargeDescriptionRequiredRule(ChargeOperationDto chargeOperationDto)
+        {
+            _chargeOperationDto = chargeOperationDto;
+        }
+
+        public bool IsValid => string.IsNullOrEmpty(_chargeOperationDto.ChargeDescription) is false;
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeDescriptionRequiredRule.cs:28]: Should we consider doing this?

```suggestion
        public bool IsValid => string.IsNullOrWhiteSpace(_chargeOperationDto.ChargeDescription) is false;
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeNameRequiredRule.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    public class ChargeNameRequiredRule : IValidationRule
+    {
+        private readonly ChargeOperationDto _chargeOperationDto;
+
+        public ChargeNameRequiredRule(ChargeOperationDto chargeOperationDto)
+        {
+            _chargeOperationDto = chargeOperationDto;
+        }
+
+        public bool IsValid => string.IsNullOrEmpty(_chargeOperationDto.ChargeName) is false;
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeNameRequiredRule.cs:28]: Again, should we do this?
```suggestion
        public bool IsValid => string.IsNullOrWhiteSpace(_chargeOperationDto.ChargeName) is false;
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs [hunk 5]
```diff
@@ -48,6 +48,8 @@ public enum ValidationRuleIdentifier
         TransparentInvoicingIsNotAllowedForFee = 34, // VR904 / D67
         ChargeResolutionCanNotBeUpdated = 35, // VR907 / D23
         RecipientRoleMustBeDdz = 36, // VR165 / E55
-        NumberOfPointsMatchTimeIntervalAndResolution = 37, // VR909 / E87
+        NumberOfPointsMatchTimeIntervalAndResolution = 37, // VR909 / E87,
+        ChargeNameIsRequired = 38, // VRXXX / E0H
+        ChargeDescriptionIsRequired = 39, // VRXXX / E0H
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs:53]: I have added them to our internal validation sheet, so please use below.

```suggestion
        ChargeNameIsRequired = 38, // VR910 / E0H
        ChargeDescriptionIsRequired = 39, // VR911 / E0H
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextTemplateMessages.cs [hunk 6]
```diff
@@ -157,6 +157,14 @@ public static class CimValidationErrorTextTemplateMessages
...
             "The number of prices received does not match the expected number of prices given the time interval and resolution provided.";
 
+        [ErrorMessageFor(ValidationRuleIdentifier.ChargeNameIsRequired)]
+        public const string ChargeNameRequiredErrorText =
+            "Charge name is missing: transaction can not be processed for document {{DocumentId}}";
+
+        [ErrorMessageFor(ValidationRuleIdentifier.ChargeDescriptionIsRequired)]
+        public const string ChargeDescriptionRequiredErrorText =
+            "Charge description is missing: transaction can not be processed for document {{DocumentId}}";
+
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextTemplateMessages.cs:167]: I suggest to keep it simple as the rejection also contains a reference to the original operation.

```suggestion
        [ErrorMessageFor(ValidationRuleIdentifier.ChargeNameIsRequired)]
        public const string ChargeNameRequiredErrorText =
            "Charge name is missing.";

        [ErrorMessageFor(ValidationRuleIdentifier.ChargeDescriptionIsRequired)]
        public const string ChargeDescriptionRequiredErrorText =
            "Charge description is missing.";

```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/ChargeCommandNullChecker.cs (hunk 2 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/ChargeCommandNullChecker.cs]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeDescriptionRequiredRuleTests.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            var dto = builder.WithDescription(null!).Build();
+            var sut = new ChargeDescriptionRequiredRule(dto);
+
+            // Assert
+            sut.IsValid.Should().BeFalse();
+        }
+
+        [Theory]
+        [InlineAutoMoqData]
+        public void IsValid_WhenChargeDescriptionIsEmpty_ShouldReturnFalse(ChargeOperationDtoBuilder builder)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeDescriptionRequiredRuleTests.cs:53]: Shall we add an `WhenChargeDescriptionIsAllWhitespaces_ShouldReturnFalse`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeNameRequiredRuleTests.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            var dto = builder.WithChargeName(null!).Build();
+            var sut = new ChargeNameRequiredRule(dto);
+
+            // Assert
+            sut.IsValid.Should().BeFalse();
+        }
+
+        [Theory]
+        [InlineAutoMoqData]
+        public void IsValid_WhenChargeNameIsEmpty_ShouldReturnFalse(ChargeOperationDtoBuilder builder)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ChargeNameRequiredRuleTests.cs:53]: Shall we add an `WhenChargeNameIsAllWhitespaces_ShouldReturnFalse`?

**Review by prtandrup — APPROVED** [review]
LGTM

**Pull request merged**